### ISSUE TITLE
on-the-fly import taxonomic information from theplantlist.org

### ISSUE THREAD

**Issue by mfrasca** [Bauble/bauble.classic#221]
- this issue (221) describes the verification of the information being inserted in the species editor window, while it is still open.
- it is related to #229 : batch correction _post facta_.
- it is also related to #225 ; sending taxonomic information to an external service — which might send a report back, again triggering corrections in the data.

@RoDuth mentioned aspects of the above issues in #163.

**Comment by RoDuth** [comment]
I was debating that myself...  Not sure as I haven't checked, but yes I most probably should have just opened the issue so thanks for doing it.

Sorry, for the next couple of weeks I am likely to be a bit short of time so I may seem like I'm dragging my feet a little or being sporadic..

**Comment by mfrasca** [comment]
I'm not so sure this what you mentioned is the same, or just similar to, the lookup functionality at http://tnrs.iplantcollaborative.org/TNRSapp.html
this one is quite good, the only missing thing is the web-api.
without a web-api, the user would need do a couple of manual steps, not so terrible according to me:
- select the things to check (this in any case)
- export with a report that will just print the genus and species epithets.
- open the above page
- paste the output into the search form
- submit list
- wait
- download results
- import the results into Bauble

what is missing?
- the export, "report that will just print the genus and species epithets".
- the import, the TNRS results.
- show a report informing the user about success and failures of the import.

**Comment by mfrasca** [comment]
ca63eae

**Comment by mfrasca** [comment]
the easiest I can imagine is: 
- you do a selection
- start the taxonomic check tool.
- it will export the relevant information into the clipboard,
- open the tnrs page.
- you need to paste the info,
- you need request the export.
- when you are offered to open the file, do I dare promise a `bauble_client`? a program that receives the file then contacts the correct running Bauble instance and passes it the information. let's say not yet.
- you save the export,
- go back to Bauble
- run the second part of the taxonomic check tool
- which does what I already wrote ca63eae, but with enough safety nets.
- then check the report.

**Comment by mfrasca** [comment]
oh, but I'm still not sure this is what @RoDuth meant with his initial comment!

**Comment by RoDuth** [comment]
I was talking about making it easy to keep our records up to date at BGCI's PlantSearch as this is one of the easiest ways to make our collections available to researchers etc.  (We already get lots of requests via this system and have worked with a couple of them) The institution data should already be in Bauble, maybe just need somewhere to save your login.  I'm sure BGCI would be happy to know there was a database option out there that could do it in one simple step... And I imagine they would be happy to help make it happen.  (may even get you a few more recommendations?) See instructions here: http://www.bgci.org/resources/plantsearchuploadinstructions/

I like where you are heading with TNRS though.  Just a different issue.  More along the lines of what I was referring to in #52  Would love to see this functionality.

**Comment by mfrasca** [comment]
oh, sorry, then I'm redirecting this issue #221 so its title matches its content, and reopen your issue with your own comments so it stays clean and clear.

**Comment by mfrasca** [comment]
I could add it to the species editor, something coming up in the message area where you are used to see warnings (and click them away).

- a new :white_check_mark: button comes next to the epithet text entry,
- you click click the :white_check_mark: button (or just the leave the epithet text entry), a request is sent to the server.
- you will get one of the following:


- "species **not** found. please check your sources before you enter this epithet."
- "species found, do you want to include this authorship?"
- additionally, "epithet has taxonomic status accepted / **synonym**".

if I am to use theplantlist.org web service, in their csv results they _mention_ the fact that an epithet is considered a synonym of something else, but you have to go online manually to get the full picture.

so, dear watchers: if you are aware of more sites offering their taxonomic information through a web service, please mention it here!

**Comment by mfrasca** [comment]
@tmyersdm, I have seen your #214 and still need find the time to have to look at what it does and how.

**Comment by RoDuth** [comment]
@mfrasca maybe not ideal for everyone but, for myself and a lot of us in Australia (and potentially NZ in the future I believe), the Atlas of Living Australia is the source I refer to the most (say 80-90% of our collection).  They seem to have some data exports and the LSID etc. (see here for an example, right hand side of screen: http://bie.ala.org.au/species/Brachychiton+bidwillii) and do have synonymy.  Also, you can download the results of a search as a csv.

I sent you an email a while back with a chain of discussion with ALA developers and others where it was mention that they were interested in assisting BGs with naming.  They could be worth contacting re: web api etc.?

The other source we use a fair bit is the [APC](http://biodiversity.org.au/nsl/services/apc) but as they are linked I tend to go to the ALA more often.  I notice that APC does have a bulk name check service also.  Have never tried it.

**Comment by RoDuth** [comment]
Also, @mfrasca.  Love the ideas above for the species editor.

Could the [LSID](https://github.com/tdwg/guid-as) help at all?

**Comment by mfrasca** [comment]
closing the day, I'm having a look at what it might mean, to correct a whole batch of species.
those having overall score 1, fine, this is useful for completing the data like authorship.

all with overall score less than 1 I would not dare send automatically to the program.
somehow the user should have a look at them and check what they want to import and what not.

the following cases seem obvious: you need correct your data:
```
1	Adiantum capillus-Veneris	0.82	Adiantum capillus-veneris	species	1	L.	http://www.theplantlist.org/tpl1.1/record/tro-26602671;http://www.tropicos.org/Name/26602671;http://plants.usda.gov/java/profile?symbol=ADCA	L.	0.11			Pteridaceae	Adiantum	1	capillus-veneris	1									Accepted	Adiantum capillus-veneris	L.	species	http://www.theplantlist.org/tpl1.1/record/tro-26602671;http://www.tropicos.org/Name/26602671;http://plants.usda.gov/java/profile?symbol=ADCA	Adiantum capillus-veneris	Pteridaceae	true	tpl;tropicos;usda	 	
8	Anthurium andranum	0.96	Anthurium andraeanum	species	0.96	Linden ex André	http://www.theplantlist.org/tpl1.1/record/kew-10633					Araceae	Anthurium	1	andraeanum	0.8									Accepted	Anthurium andraeanum	Linden ex André	species	http://www.theplantlist.org/tpl1.1/record/kew-10633	Anthurium andraeanum	Araceae	true	tpl	 	
15	Astrocaryua malybo	0.99	Astrocaryum malybo	species	0.99	H.Karst.	http://www.theplantlist.org/tpl1.1/record/kew-17547					Arecaceae	Astrocaryum	0.91	malybo	1									Accepted	Astrocaryum malybo	H.Karst.	species	http://www.theplantlist.org/tpl1.1/record/kew-17547	Astrocaryum malybo	Arecaceae	true	tpl	 	
```

but number 15 becomes less obvious considering there is already an _Astrocaryum malybo_ in the database. so you need to edit things by hand, even for a close-to-perfect match of 0.99.

**Comment by mfrasca** [comment]
@RoDuth , no, not ideal, the http://biodiversity.org.au/taxon/ site.
it does offer quite complete information, but we really need something that covers us worldwide.
I tried just at random 4 species from the Brako & Zarucchi list (Catalogue ... of Peru) and it was four 404.

**Comment by mfrasca** [comment]
`http://www.theplantlist.org/tpl1.1/search?q=%(sp.genus.epithet)s %(sp.epithet)s&csv=true`
when a row is marked as Synonym, its last field is the accepted ID, which we can retrieve similarly

for example, I look for `Cyrtidium stumpflei`:
```
kew-54471,A,Orchidaceae,,Cyrtidium,,"stumpflei",,"","Garay",Synonym,,H,WCSP,54471,75438-2,"Orquideologia","4: 6","","1969",kew-54465
```
I have to do a second query for `kew-54465` and get:
```
kew-54465,A,Orchidaceae,,Cyrtidiorchis,,"stumpflei",,"","(Garay) Rauschert",Accepted,,H,WCSP,54465,910482-1,"Taxon","31: 560","","1982",
```

**Comment by RoDuth** [comment]
My understanding is that the LSID is a direct link to the data... regardless of the name.  Kind of like an accession number for the taxon, (although it can be used in other ways).  It also contains the resource that it came from. e.g. `urn:lsid:biodiversity.org.au:apni.taxon:754049` points to *Brachychiton bidwillii* as defined by APNI (at biodiversity.org.au) and `urn:lsid:ipni.org:names:530017-1` points to *Aloe vera* as defined by IPNI (at ipni.org).  Kind of like your `kew-54465` above.

@tmyersdn is the one who first suggested that it would be handy to keep in the database and may be able to exand on this??

I put APC and ALA forward because I figured that if it was easy to work with and you have contacts for the developers etc. to ask any questions that it may be a good test case?

**Comment by mfrasca** [comment]
ah! this is what I missed: **it would be handy to keep [the LSID] in the database**.
ok. sounds like a new issue. feel free to open! :beetle:

**Comment by mfrasca** [comment]
string distances will be relevant. see for example [this question](http://stackoverflow.com/questions/1471153/string-similarity-metrics-in-python)

**Comment by mfrasca** [comment]
the logic is all in place and you can see what it does in the logging window, if you open a terminal and `tail -f ~/.bauble/bauble.log`.
now I have to write the part showing the information in self.view.add_message_box and handling the response.

**Comment by mfrasca** [comment]
still to do: it does not report when a query ends (and your data is fine), nor when it does not get any reply from the server, nor does it check discrepancies in authorship, and I'm not sure what to do with omonyms from different authors.
![screenshot from 2015-12-23 12 04 05](https://cloud.githubusercontent.com/assets/122869/11981069/56b954de-a96d-11e5-95c4-760affe8432b.png)

**Comment by mfrasca** [comment]
oh, I'm not so sure how it should behave with infraspecific information.
I have _Bauhinia splendens_, which is a synonym of _Bauhinia guianensis_ var. _splendens_ (Kunth) Amshoff.
the code I've implemented ignores everything which is not at rank species.
so it will find the information for _Bauhinia splendens_, see it's marked as synonym, look for the accepted name, discard all varieties and formae, see that nothing matches, and conclude "no match found".
I think I prefer leaving this for when I finally find the energies to unit-test this presenter.

**Comment by mfrasca** [comment]
I check »_Annona squamosa_ L.«, the program finds several omonyms, and chooses randomly among them, so it can return »_Annona squamosa_ Delile«, offer correction but also offer to add »_Annona squamosa_ L.« as the accepted taxon.
expected result is "fine match, well done".

**Comment by mfrasca** [comment]
what happens here with _Brownea ariza_?
![screenshot from 2015-12-23 17 09 17](https://cloud.githubusercontent.com/assets/122869/11985945/f1094c04-a997-11e5-8c28-4341e2aa23cd.png)

**Comment by mfrasca** [comment]
the above two problems are solved in `master` and will be fine in 1.0.54.

**Comment by mfrasca** [comment]
forgot to close
